src/armature/FillVolume.ts

```diff
@@ -0,0 +1,143 @@
+import { AABB } from '../geometry/BakedGeometry';
+import { worldSpaceAABB } from '../utils/aabb';
+import { Cost, GeneratorInstance, SpawnPoint } from './Generator';
+import { Model } from './Model';
+import { GeometryNode, Node } from './Node';
+
+import { vec4 } from 'gl-matrix';
+import { range } from 'lodash';
+
+type Grid = { [key: string]: true };
+
+/**
+ * Creates a cost function based on how much of a target volume a shape fills.
+ */
+export class FillVolume {
+    private gridCache: Map<Node, Grid> = new Map<Node, Grid>();
+    private cellSize: number;
+    private targetCoords: Grid = {};
+
+    /**
+     * @param {Model} targetModel The model whose shape we try to fill.
+     * @param {number} cellSize How big each cell in the volume grid should be.
+     * @returns {CostFn} The resulting cost function.
+     */
+    constructor(targetModel: Model, cellSize: number) {
+        this.cellSize = cellSize;
+
+        // For each node in the target model, find its bounding box
+        targetModel.nodes.forEach((n: Node) =>
+            n.geometryCallback((node: GeometryNode) => {
+                this.pointsInAABB(worldSpaceAABB(node, node.geometry.aabb)).forEach(
+                    (point: string) => (this.targetCoords[point] = true)
+                );
+            })
+        );
+    }
+
+    public getCost(instance: GeneratorInstance, added: Node[]): Cost {
+        // We will cache grids based on the last node that was added to a model. Since nodes are
+        // added to the end of a model's node list, if n nodes are new in the current model
+        // compared to its parent, then the parent grid will be indexed by the nth-from-last
+        // node in the current model's node list.
+        const parentGrid = this.gridCache.get(
+            instance.getModel().nodes[instance.getModel().nodes.length - 1 - added.length]
+        );
+
+        // If the parent grid does exist, we want to start from the same grid as the
+        // parent, and then add to it
+        const grid = parentGrid === undefined ? {} : { ...parentGrid };
+
+        let incrementalCost = 0;
+        let heuristicCost = 0;
+
+        // For each point in the new added geometry, see if it fills a point in the target shape
+        // that wasn't previously filled, and make the current model cost less accordingly
+        added.forEach((n: Node) =>
+            n.geometryCallback((node: GeometryNode) =>
+                this.addAABBToGrid(
+                    worldSpaceAABB(node, node.geometry.aabb),
+                    grid,
+                    (point: string) =>
+                        // If this point was in the target region, reduce the cost
+                        (incrementalCost +=
```

Comment:
is this the autoformatter's doing? 🤕 lol

Reply:
yep :')